File: cotgenerator/src/main/java/com/jon/cot/generator/utils/NetworkHelper.java
package com.jon.cot.generator.utils;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import timber.log.Timber;

public class NetworkHelper {
    private NetworkHelper() { }

    public static List<NetworkInterface> getValidInterfaces() throws SocketException {
        List<NetworkInterface> interfaces = new ArrayList<>();
        for (NetworkInterface ni : Collections.list(NetworkInterface.getNetworkInterfaces())) {
            if (ni.isUp() && interfaceHasIpv4Address(ni)) {
                interfaces.add(ni);
                Timber.d("%s is up", ni.getName());
            }
        }
        return interfaces;
    }

    public static InetAddress getAddressFromInterface(NetworkInterface ni) {
        for (InterfaceAddress interfaceAddress : ni.getInterfaceAddresses()) {
            if (interfaceAddress.getAddress() instanceof Inet4Address) {
                return interfaceAddress.getAddress();
            }
        }
        return null;
    }

    private static boolean interfaceHasIpv4Address(NetworkInterface ni) {
        return getAddressFromInterface(ni) != null;
    }
}

File: common/src/main/java/com/jon/common/utils/NetworkHelper.kt
package com.jon.common.utils

import java.net.Inet4Address
import java.net.InetAddress
import java.net.NetworkInterface

class NetworkHelper {
    fun getValidInterfaces(): List<NetworkInterface> {
        return NetworkInterface.getNetworkInterfaces()
                .asSequence()
                .filter { it.isUp && interfaceHasIpv4Address(it) }
                .toList()
    }

    fun getAddressFromInterface(ni: NetworkInterface): InetAddress? {
        return ni.interfaceAddresses.map { it.address }.firstOrNull { it is Inet4Address }
    }

    private fun interfaceHasIpv4Address(ni: NetworkInterface): Boolean {
        return getAddressFromInterface(ni) != null
    }
}
